File: POTD/DP/Digit DP/Python/3753. Total Waviness of Numbers in Range II.py
class Solution:
    def func(self, num: int) -> int:
        if num < 100:
            return 0
        
        s = str(num)
        n = len(s)
        dp = {}

        def solve(pos: int, prev_prev: int, prev: int, is_leading_zero: bool, is_part_equal: bool) -> tuple[int, int]:
            if pos >= n:
                return (1, 0)
            
            state = (pos, prev_prev, prev, is_leading_zero)

            if not is_part_equal and state in dp:
                return dp[state]
            
            total_numbers = 0
            total_wave_score = 0

            digit_limit = int(s[pos]) if is_part_equal else 9

            for digit in range(digit_limit + 1):
                is_new_leading_zero = is_leading_zero and (digit == 0)
                new_prev_prev = prev
                new_prev = -1 if is_new_leading_zero else digit

                rem_total_count, rem_total_score = solve(
                    pos + 1,
                    new_prev_prev,
                    new_prev,
                    is_new_leading_zero,
                    is_part_equal and (digit == digit_limit)
                )

                if not is_new_leading_zero and prev >= 0 and prev_prev >= 0:
                    is_peak = (prev > digit) and (prev > prev_prev)
                    is_valley = (prev < digit) and (prev < prev_prev)

                    if is_peak or is_valley:
                        total_wave_score += rem_total_count
                
                total_numbers += rem_total_count
                total_wave_score += rem_total_score

            if not is_part_equal:
                dp[state] = (total_numbers, total_wave_score)
            
            return (total_numbers, total_wave_score)

        _, total_wave_score = solve(0, -1, -1, True, True)

        return total_wave_score

    def totalWaviness(self, num1: int, num2: int) -> int:
        return self.func(num2) - self.func(num1 - 1)

File: POTD/DP/Digit DP/CPP/3753. Total Waviness of Numbers in Range II.cpp
using ll = long long;
class Solution {
private:
    ll dpCount[18][12][12][2];
    ll dpScore[18][12][12][2];

    pair<ll, ll> solve(int pos, int prevPrv, int prev, bool isLeadingZero,
                       bool isPartEqual, string& s) {
        if (pos >= s.length()) {
            return {1, 0};
        }

        int ppIdx = prevPrv + 1;
        int pIdx = prev + 1;
        int leadIdx = isLeadingZero ? 1 : 0;

        if (!isPartEqual && dpCount[pos][ppIdx][pIdx][leadIdx] != -1) {
            return {dpCount[pos][ppIdx][pIdx][leadIdx],
                    dpScore[pos][ppIdx][pIdx][leadIdx]};
        }
        ll totalNumbers = 0;
        ll totalWaveScore = 0;

        int digitLimit = isPartEqual ? (s[pos] - '0') : 9;

        for (int digit = 0; digit <= digitLimit; digit++) {
            bool isNewLeadingZero = isLeadingZero && (digit == 0);
            int newPrvPrv = prev;
            int newPrev = isNewLeadingZero ? -1 : digit;

            auto [remTotalCount, remTotalScore] =
                solve(pos + 1, newPrvPrv, newPrev, isNewLeadingZero,
                      isPartEqual && (digit == digitLimit), s);

            if (!isNewLeadingZero && prev >= 0 && prevPrv >= 0) {
                bool isPeak = (prev > digit && prev > prevPrv);
                bool isVall = (prev < digit && prev < prevPrv);

                if (isPeak || isVall) {
                    totalWaveScore += (1 * remTotalCount);
                }
            }

            totalNumbers += remTotalCount;
            totalWaveScore += remTotalScore;
        }

        if (!isPartEqual) {
            dpCount[pos][ppIdx][pIdx][leadIdx] = totalNumbers;
            dpScore[pos][ppIdx][pIdx][leadIdx] = totalWaveScore;
        }
        return {totalNumbers, totalWaveScore};
    }
    ll func(long long num) {
        if (num < 100) {
            return 0;
        }

        string s = to_string(num);
        memset(dpCount, -1, sizeof(dpCount));
        memset(dpScore, -1, sizeof(dpScore));
        auto [totalNumbers, totalWaveScore] = solve(0, -1, -1, true, true, s);

        return totalWaveScore;
    }

public:
    long long totalWaviness(long long num1, long long num2) {

        
        return func(num2) - func(num1 - 1);
    }
};
